## Modified K-Nearest Neighbor

In [139]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import NearestNeighbors # NearestNeighbors Algorithms

In [140]:
data = pd.read_csv('waterQuality1.csv')
pd.set_option('display.max_columns', None)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7999 entries, 0 to 7998
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   aluminium    7999 non-null   float64
 1   ammonia      7999 non-null   object 
 2   arsenic      7999 non-null   float64
 3   barium       7999 non-null   float64
 4   cadmium      7999 non-null   float64
 5   chloramine   7999 non-null   float64
 6   chromium     7999 non-null   float64
 7   copper       7999 non-null   float64
 8   flouride     7999 non-null   float64
 9   bacteria     7999 non-null   float64
 10  viruses      7999 non-null   float64
 11  lead         7999 non-null   float64
 12  nitrates     7999 non-null   float64
 13  nitrites     7999 non-null   float64
 14  mercury      7999 non-null   float64
 15  perchlorate  7999 non-null   float64
 16  radium       7999 non-null   float64
 17  selenium     7999 non-null   float64
 18  silver       7999 non-null   float64
 19  uraniu

In [141]:
# menampilkan data awal 
data.head(10)

aluminium ammonia  arsenic  barium  cadmium  chloramine  chromium  copper   
0       1.65    9.08     0.04    2.85    0.007        0.35      0.83    0.17  \
1       2.32   21.16     0.01    3.31    0.002        5.28      0.68    0.66   
2       1.01   14.02     0.04    0.58    0.008        4.24      0.53    0.02   
3       1.36   11.33     0.04    2.96    0.001        7.23      0.03    1.66   
4       0.92   24.33     0.03    0.20    0.006        2.67      0.69    0.57   
5       0.94   14.47     0.03    2.88    0.003        0.80      0.43    1.38   
6       2.36     5.6     0.01    1.35    0.004        1.28      0.62    1.88   
7       3.93   19.87     0.04    0.66    0.001        6.22      0.10    1.86   
8       0.60   24.58     0.01    0.71    0.005        3.14      0.77    1.45   
9       0.22   16.76     0.02    1.37    0.007        6.40      0.49    0.82   

   flouride  bacteria  viruses   lead  nitrates  nitrites  mercury   
0      0.05      0.20    0.000  0.054     16.08      1.13    0.007  \
1      0.90      0.65    0.650  0.100      2.01      1.93    0.003   
2      0.99      0.05    0.003  0.078     14.16      1.11    0.006   
3      1.08      0.71    0.710  0.016      1.41      1.29    0.004   
4      0.61      0.13    0.001  0.117      6.74      1.11    0.003   
5      0.11      0.67    0.670  0.135      9.75      1.89    0.006   
6      0.33      0.13    0.007  0.021     18.60      1.78    0.007   
7      0.86      0.16    0.005  0.197     13.65      1.81    0.001   
8      0.98      0.35    0.002  0.167     14.66      1.84    0.004   
9      1.24      0.83    0.830  0.109      4.79      1.46    0.010   

   perchlorate  radium  selenium  silver  uranium is_safe  
0        37.75    6.78      0.08    0.34     0.02       1  
1        32.26    3.21      0.08    0.27     0.05       1  
2        50.28    7.07      0.07    0.44     0.01       0  
3         9.12    1.72      0.02    0.45     0.05       1  
4        16.90    2.41      0.02    0.06     0.02       1  
5        27.17    5.42      0.08    0.19     0.02       1  
6        45.34    2.84      0.10    0.24     0.08       0  
7        53.35    7.24      0.08    0.08     0.07       0  
8        23.43    4.99      0.08    0.25     0.08       1  
9        30.42    0.08      0.03    0.31     0.01       1

### Data Preprocessing

In [142]:
# mengecek data duplikat
duplicates = data.duplicated()
data.drop_duplicates(inplace=True)

# mengecek data kosong
empty = data.isnull()
data.dropna(inplace=True)

data.head(10)

aluminium ammonia  arsenic  barium  cadmium  chloramine  chromium  copper   
0       1.65    9.08     0.04    2.85    0.007        0.35      0.83    0.17  \
1       2.32   21.16     0.01    3.31    0.002        5.28      0.68    0.66   
2       1.01   14.02     0.04    0.58    0.008        4.24      0.53    0.02   
3       1.36   11.33     0.04    2.96    0.001        7.23      0.03    1.66   
4       0.92   24.33     0.03    0.20    0.006        2.67      0.69    0.57   
5       0.94   14.47     0.03    2.88    0.003        0.80      0.43    1.38   
6       2.36     5.6     0.01    1.35    0.004        1.28      0.62    1.88   
7       3.93   19.87     0.04    0.66    0.001        6.22      0.10    1.86   
8       0.60   24.58     0.01    0.71    0.005        3.14      0.77    1.45   
9       0.22   16.76     0.02    1.37    0.007        6.40      0.49    0.82   

   flouride  bacteria  viruses   lead  nitrates  nitrites  mercury   
0      0.05      0.20    0.000  0.054     16.08      1.13    0.007  \
1      0.90      0.65    0.650  0.100      2.01      1.93    0.003   
2      0.99      0.05    0.003  0.078     14.16      1.11    0.006   
3      1.08      0.71    0.710  0.016      1.41      1.29    0.004   
4      0.61      0.13    0.001  0.117      6.74      1.11    0.003   
5      0.11      0.67    0.670  0.135      9.75      1.89    0.006   
6      0.33      0.13    0.007  0.021     18.60      1.78    0.007   
7      0.86      0.16    0.005  0.197     13.65      1.81    0.001   
8      0.98      0.35    0.002  0.167     14.66      1.84    0.004   
9      1.24      0.83    0.830  0.109      4.79      1.46    0.010   

   perchlorate  radium  selenium  silver  uranium is_safe  
0        37.75    6.78      0.08    0.34     0.02       1  
1        32.26    3.21      0.08    0.27     0.05       1  
2        50.28    7.07      0.07    0.44     0.01       0  
3         9.12    1.72      0.02    0.45     0.05       1  
4        16.90    2.41      0.02    0.06     0.02       1  
5        27.17    5.42      0.08    0.19     0.02       1  
6        45.34    2.84      0.10    0.24     0.08       0  
7        53.35    7.24      0.08    0.08     0.07       0  
8        23.43    4.99      0.08    0.25     0.08       1  
9        30.42    0.08      0.03    0.31     0.01       1

In [143]:
# cek data tidak relevan
# Ganti '#NUM!' dengan NaN
data = data.replace('#NUM!', np.nan)

# Pastikan untuk mengonversi semua kolom ke tipe yang benar
data = data.astype(float)

# Sekarang Anda dapat menghitung korelasinya
correlations = data.corr()['is_safe'].sort_values()

# fitur dengan korelasi absolut kurang dari ambang batas 1,0001 sebagai tidak relevan
irrelevant_features = correlations[np.abs(correlations) < 1.0001]

print(irrelevant_features)

cadmium       -0.256034
arsenic       -0.123362
viruses       -0.097040
uranium       -0.075619
nitrates      -0.072100
mercury       -0.036817
selenium      -0.030923
ammonia       -0.022919
bacteria      -0.022077
lead          -0.009970
flouride       0.006550
copper         0.029498
nitrites       0.046986
radium         0.064743
perchlorate    0.075735
barium         0.090858
silver         0.102813
chromium       0.182306
chloramine     0.186704
aluminium      0.333974
is_safe        1.000000
Name: is_safe, dtype: float64


### Data Transformation

In [144]:
# Transformasi data dengan normalisasi min-max
# Kolom yang akan di normalisasi
columns = ['aluminium', 'ammonia', 'arsenic', 'barium', 'cadmium', 'chloramine', 'chromium', 'copper', 'flouride', 'bacteria', 'viruses', 'lead', 'nitrates', 'nitrites', 'mercury', 'perchlorate', 'radium', 'selenium', 'silver', 'uranium']
columns1 = ['aluminium', 'ammonia', 'arsenic', 'barium', 'cadmium', 'chloramine', 'chromium', 'copper', 'flouride', 'bacteria', 'viruses', 'lead', 'nitrates', 'nitrites', 'mercury', 'perchlorate', 'radium', 'selenium', 'silver', 'uranium','is_safe']

# Melakukan normalisasi Min-Max hanya untuk 10 data pertama
num_data = 10
normalized = (data[columns].head(num_data) - data[columns].min()) / (data[columns].max() - data[columns].min())
normalized1 = (data[columns1].head(num_data) - data[columns1].min()) / (data[columns1].max() - data[columns1].min())

# Tampilkan hasil normalisasi
print(normalized)

   aluminium   ammonia   arsenic    barium   cadmium  chloramine  chromium   
0   0.326733  0.306150  0.038095  0.576923  0.053846    0.040323  0.922222  \
1   0.459406  0.709893  0.009524  0.670040  0.015385    0.608295  0.755556   
2   0.200000  0.471257  0.038095  0.117409  0.061538    0.488479  0.588889   
3   0.269307  0.381350  0.038095  0.599190  0.007692    0.832949  0.033333   
4   0.182178  0.815842  0.028571  0.040486  0.046154    0.307604  0.766667   
5   0.186139  0.486297  0.028571  0.582996  0.023077    0.092166  0.477778   
6   0.467327  0.189840  0.009524  0.273279  0.030769    0.147465  0.688889   
7   0.778218  0.666778  0.038095  0.133603  0.007692    0.716590  0.111111   
8   0.118812  0.824198  0.009524  0.143725  0.038462    0.361751  0.855556   
9   0.043564  0.562834  0.019048  0.277328  0.053846    0.737327  0.544444   

   copper  flouride  bacteria  viruses   lead  nitrates  nitrites  mercury   
0   0.085  0.033333      0.20    0.000  0.270  0.810893  0.3856

### Menghitung Jarak Antar Data Latih Menggunakan Euclidean Distance

In [145]:
# Konversi Pandas DataFrame ke array NumPy
data_array = normalized.to_numpy()

# Atur ukuran batch
batch_size = 10

# Get the number of rows in the data
num_rows = data_array.shape[0]

# Dapatkan jumlah baris dalam data
distances1 = np.zeros((num_rows, num_rows))

# Hitung jarak dalam batch
for i in range(0, num_rows, batch_size):
    start_idx = i
    end_idx = min(i + batch_size, num_rows)
    
    # Menghitung jarak Euclidean antar data
    distances1[start_idx:end_idx, :] = np.linalg.norm(data_array[start_idx:end_idx, np.newaxis, :] - data_array, axis=2)

# Cetak matriks jarak
print(distances1)

[[0.         1.63636365 1.06226913 2.24616242 1.55662491 1.33573804
  1.35638469 2.06035929 1.57531978 2.04255022]
 [1.63636365 0.         1.60057941 1.37903593 1.47403045 1.17369774
  1.75164482 1.74044973 1.36391849 1.29610167]
 [1.06226913 1.60057941 0.         2.05101073 1.44684849 1.63978128
  1.6442515  1.77756153 1.46617663 1.76830533]
 [2.24616242 1.37903593 2.05101073 0.         1.90865551 1.69264736
  2.08741729 2.1695345  1.96966491 1.31894986]
 [1.55662491 1.47403045 1.44684849 1.90865551 0.         1.52919874
  1.80707385 1.79236456 1.29918678 1.63293111]
 [1.33573804 1.17369774 1.63978128 1.69264736 1.52919874 0.
  1.53361731 1.76194882 1.41659245 1.50631878]
 [1.35638469 1.75164482 1.6442515  2.08741729 1.80707385 1.53361731
  0.         1.67367728 1.34378081 2.14023778]
 [2.06035929 1.74044973 1.77756153 2.1695345  1.79236456 1.76194882
  1.67367728 0.         1.34057252 2.30185933]
 [1.57531978 1.36391849 1.46617663 1.96966491 1.29918678 1.41659245
  1.34378081 1.34057

### Hasil Euclidean Distance Dari Jarak Antar Data Latih ###
[[0.         1.63636365 1.06226913 2.24616242 1.55662491 1.33573804 1.35638469 2.06035929 1.57531978 2.04255022]

 [1.63636365 0.         1.60057941 1.37903593 1.47403045 1.17369774 1.75164482 1.74044973 1.36391849 1.29610167]

 [1.06226913 1.60057941 0.         2.05101073 1.44684849 1.63978128 1.6442515  1.77756153 1.46617663 1.76830533]

 [2.24616242 1.37903593 2.05101073 0.         1.90865551 1.69264736 2.08741729 2.1695345  1.96966491 1.31894986]

 [1.55662491 1.47403045 1.44684849 1.90865551 0.         1.52919874 1.80707385 1.79236456 1.29918678 1.63293111]

 [1.33573804 1.17369774 1.63978128 1.69264736 1.52919874 0.         1.53361731 1.76194882 1.41659245 1.50631878]

 [1.35638469 1.75164482 1.6442515  2.08741729 1.80707385 1.53361731 0.         1.67367728 1.34378081 2.14023778]

 [2.06035929 1.74044973 1.77756153 2.1695345  1.79236456 1.76194882 1.67367728 0.         1.34057252 2.30185933]

 [1.57531978 1.36391849 1.46617663 1.96966491 1.29918678 1.41659245 1.34378081 1.34057252 0.         1.82878264]
 
 [2.04255022 1.29610167 1.76830533 1.31894986 1.63293111 1.50631878 2.14023778 2.30185933 1.82878264 0.        ]]


### Mengitung Validitas Data

In [146]:
# Data target (kelas) yang sesuai dengan urutan matriks jarak
target_classes = data['is_safe'].head(10)

# Ambil hanya 100 data untuk dihitung validitasnya
num_data = 10
distances_subset = distances1[:num_data]
target_classes_subset = target_classes[:num_data]

# Number of neighbors (K)
k = 5

# Inisialisasi array untuk menyimpan validitas per fitur
feature_validities = []

# Loop untuk setiap fitur
for feature_index in range(distances_subset.shape[1]):  # Menggunakan distances_subset.shape[1]
    feature_distances = distances_subset[:, feature_index]
    
    # Urutkan indeks data berdasarkan jarak fitur
    sorted_indices = np.argsort(feature_distances)
    
    # Ambil hanya K data terdekat untuk setiap data
    closest_indices = sorted_indices[:k]
    
    # Hitung validitas untuk setiap data
    feature_validity = np.sum(target_classes_subset.values[closest_indices] == target_classes_subset.values[i]) / num_data
    
    # Simpan validitas fitur ke dalam array
    feature_validities.append(feature_validity)

# Print hasil validitas per fitur
for feature_index, validity in enumerate(feature_validities):
    print(f"Validitas fitur-{feature_index + 1}: {validity}")

Validitas fitur-1: 0.3
Validitas fitur-2: 0.5
Validitas fitur-3: 0.4
Validitas fitur-4: 0.5
Validitas fitur-5: 0.4
Validitas fitur-6: 0.5
Validitas fitur-7: 0.3
Validitas fitur-8: 0.3
Validitas fitur-9: 0.3
Validitas fitur-10: 0.5


### Hitung jarak data uji dan setiap baris dalam data latih

In [147]:
# Contoh data uji
data_uji = {
    'aluminium': 0.25,
    'ammonia': 0.45,
    'arsenic': 0.05,
    'barium': 0.35,
    'cadmium': 0.15,
    'chloramine': 0.55,
    'chromium': 0.65,
    'copper': 0.75,
    'flouride': 0.85,
    'bacteria': 0.95,
    'viruses': 0.05,
    'lead': 0.15,
    'nitrates': 0.25,
    'nitrites': 0.35,
    'mercury': 0.45,
    'perchlorate': 0.55,
    'radium': 0.65,
    'selenium': 0.75,
    'silver': 0.85,
    'uranium': 0.95
    # 'is_safe': 1.0  # Nilai ini akan kita prediksi
}
data_uji = pd.DataFrame([data_uji])
data_uji


aluminium  ammonia  arsenic  barium  cadmium  chloramine  chromium  copper   
0       0.25     0.45     0.05    0.35     0.15        0.55      0.65    0.75  \

   flouride  bacteria  viruses  lead  nitrates  nitrites  mercury   
0      0.85      0.95     0.05  0.15      0.25      0.35     0.45  \

   perchlorate  radium  selenium  silver  uranium  
0         0.55    0.65      0.75    0.85     0.95

In [148]:
# Hitung jarak Euclidean antara data uji dan setiap baris dalam data latih
distances2 = np.linalg.norm(normalized[columns].to_numpy() - data_uji[columns].to_numpy(), axis=1)

print(distances2)

[1.75926626 1.23100761 1.61551658 1.38562875 1.78451928 1.60998273
 1.53092682 1.79410526 1.27922044 1.64534863]


### Hitung Bobot Data

In [151]:
# Nilai alpha
alpha = 0.5

# Hitung bobot untuk setiap data latih
weights = validity * (1 / (distances2 + alpha))

# Print hasil bobot
print("Bobot :",end="")
print(weights)

Bobot :[0.22131079 0.2888491  0.23634889 0.26516354 0.21886443 0.23696876
 0.24619302 0.21794989 0.28102195 0.23306235]


### Hasil Prediksi Kelas

In [157]:
# Menambahkan kolom bobot ke normalized
normalized1['weights'] = weights

# Mengurutkan data berdasarkan bobot
sorted_data = normalized1.sort_values(by='weights')

# Mengekstrak K tetangga terdekat (K=5)
k = 5
k_neighbors = sorted_data.head(k)

# Menentukan kelas mayoritas dari K tetangga terdekat
prediction = k_neighbors['is_safe'].mode().values[0]

# Print hasil prediksi
print("Prediksi Kelas: ", prediction)

Prediksi Kelas:  1.0
